### DISCUSSION TITLE
Overridable core features

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#779] [Proposal]
### Body

# Summary

Add ways for users to override core features, likely using integrations

# Background & Motivation

Some core Astro features are kept internal for good reasons (eg. less maintenance) but it can become a bit limiting in more advanced cases. For example, having another routing strategy or even another directory for pages (eg. app like in nextjs).

# Goals

- Make Astro core features overridable hence more configurable
- Do not allow this extensibility through configuration

# Example

I think a great example of this is Nuxt. Nuxt is built-on top of modules, everything (maybe not but a lot of stuff at least) is a module with an id. The great thing is that 3rd party modules provided with `@nuxt/kit` can override existing modules if they are defined afterwards. I can't remember what module does this but I remember one override the `pages` module

**Comment by matthewp** [comment]
I would suggest opening a proposal for the feature that you'd like to be overridden. As you note, we already do have this to some degree with integrations. I don't think "making everything overriddeable" is really achievable unless the project starts from scratch with that as a primary goal. But, given some specific features that you'd like to be, we can discuss possible solutions.

**Reply by florian-lefebvre** [reply]
So far, the only usecase I can see is routing but I'm curious to see what others may need
